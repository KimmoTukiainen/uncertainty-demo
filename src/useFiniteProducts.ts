import useFakeApiCall from './useFakeApiCall';
import productData from './products';

import { Product } from './useProducts';

const normalizeProduct = (product: Product): FiniteProduct => ({
  ...product,
  name: product.name,
  price: {
    amount: product.price?.amount ?? null,
    unit: product.price?.unit ?? 'â‚¬',
  },
});

type FinitePrice = {
  amount: number | null;
  unit: string;
};

export type FiniteProduct = {
  name: string;
  price: FinitePrice;
};

type Success = {
  name: 'success';
  products: FiniteProduct[];
};

type Failure = {
  name: 'failure';
  error: Error;
};

type Loading = {
  name: 'loading';
};

export type FiniteApiResult = Success | Failure | Loading;

export default (): FiniteApiResult => {
  const { loading, result: products, error } = useFakeApiCall(productData);

  if (loading) {
    return {
      name: 'loading',
    };
  }

  if (products) {
    return {
      name: 'success',
      products: products.map(normalizeProduct),
    };
  }

  return {
    name: 'failure',
    error: error || new Error('No data received'),
  };
};
