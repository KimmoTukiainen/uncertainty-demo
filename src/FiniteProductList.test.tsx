import { RenderResult, render } from '@testing-library/react';
import React from 'react';

import FiniteProducList from './FiniteProductList';
import useFiniteProducts from './useFiniteProducts';

jest.mock('./useFiniteProducts');

describe('ProductList', () => {
  let result: RenderResult;
  let useProductsMock = useFiniteProducts as jest.Mock;

  it('shows a loader', () => {
    useProductsMock.mockImplementationOnce(() => ({
      name: 'loading',
    }));
    result = render(<FiniteProducList />);

    expect(result.getByTestId('loading')).toMatchSnapshot();
  });

  it('shows an error', () => {
    useProductsMock.mockImplementationOnce(() => ({
      name: 'failure',
      error: new Error('some error'),
    }));
    result = render(<FiniteProducList />);

    expect(result.getByTestId('error')).toMatchSnapshot();
  });

  it('shows a list of products', () => {
    useProductsMock.mockImplementationOnce(() => ({
      name: 'success',
      products: [
        {
          name: 'some name',
          price: {
            amount: 10.5,
            unit: 'â‚¬',
          },
        },
      ],
    }));
    result = render(<FiniteProducList />);

    expect(result.getByTestId('success')).toMatchSnapshot();
  });
});
