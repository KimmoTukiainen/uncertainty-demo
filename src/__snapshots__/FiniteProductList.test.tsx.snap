// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList shows a list of products 1`] = `
<div
  data-testid="success"
>
  <p>
    <span>
      some name
    </span>
    <span>
      10.5
      â‚¬
    </span>
  </p>
</div>
`;

exports[`ProductList shows a loader 1`] = `
<p
  data-testid="loading"
>
  loading
</p>
`;

exports[`ProductList shows an error 1`] = `
<p
  data-testid="error"
>
  some error
</p>
`;
